File: src/main/java/org/algohub/engine/compiler/java/MemoryJavaFileManager.java
package org.algohub.engine.compiler.java;

import javax.tools.*;
import javax.tools.JavaFileObject.Kind;
import java.io.ByteArrayOutputStream;
import java.io.FilterOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;

/**
 * JavaFileManager that keeps compiled .class bytes in memory.
 */
final class MemoryJavaFileManager extends ForwardingJavaFileManager<JavaFileManager> {

    private static final String JAVA_SOURCE_FILE_EXT = ".java";

    private Map<String, byte[]> classBytes;

    MemoryJavaFileManager(JavaFileManager fileManager) {
        super(fileManager);
        classBytes = new HashMap<>();
    }

    static URI toUri(String name) {
        final StringBuilder newUri = new StringBuilder();
        newUri.append("mfm:///");
        newUri.append(name.replace('.', '/'));
        if (name.endsWith(JAVA_SOURCE_FILE_EXT)) {
            newUri.replace(newUri.length() - JAVA_SOURCE_FILE_EXT.length(), newUri.length(), JAVA_SOURCE_FILE_EXT);
        }
        return URI.create(newUri.toString());
    }

    Map<String, byte[]> getClassBytes() {
        return classBytes;
    }

    @Override
    public void close() throws IOException {
        classBytes = null;
    }

    @Override
    public void flush() throws IOException {
        // in memory, no action
    }

    @Override
    public JavaFileObject getJavaFileForOutput(JavaFileManager.Location location, String className,
                                               Kind kind, FileObject sibling) throws IOException {
        if (kind == Kind.CLASS) {
            return new ClassOutputBuffer(className);
        } else {
            return super.getJavaFileForOutput(location, className, kind, sibling);
        }
    }

    /**
     * A file object that stores Java bytecode into the classBytes map.
     */
    private class ClassOutputBuffer extends SimpleJavaFileObject {
        private String name;

        ClassOutputBuffer(String name) {
            super(MemoryJavaFileManager.toUri(name), Kind.CLASS);
            this.name = name;
        }

        @Override
        public OutputStream openOutputStream() {
            return new FilterOutputStream(new ByteArrayOutputStream()) {
                @Override
                public void close() throws IOException {
                    out.close();
                    ByteArrayOutputStream bos = (ByteArrayOutputStream) out;
                    classBytes.put(name, bos.toByteArray());
                }
            };
        }
    }
}

File: src/main/kotlin/org/algohub/engine/compile/MemoryJavaFileManager.kt
package org.algohub.engine.compile

import java.io.ByteArrayOutputStream
import java.io.FilterOutputStream
import java.io.OutputStream
import java.net.URI
import java.util.*
import javax.tools.*
import javax.tools.JavaFileObject.Kind

internal class MemoryJavaFileManager(fileManager: JavaFileManager) : ForwardingJavaFileManager<JavaFileManager>(fileManager) {

    var classBytes: MutableMap<String, ByteArray?>? = HashMap()

    override fun close() {
        classBytes = null
    }

    override fun getJavaFileForOutput(location: JavaFileManager.Location, className: String,
                                      kind: Kind, sibling: FileObject): JavaFileObject {
        return if (kind == Kind.CLASS) {
            ClassOutputBuffer(className)
        } else {
            super.getJavaFileForOutput(location, className, kind, sibling)
        }
    }

    private inner class ClassOutputBuffer(val className: String)
        : SimpleJavaFileObject(Companion.toUri(className), Kind.CLASS) {

        override fun openOutputStream(): OutputStream {
            return object : FilterOutputStream(ByteArrayOutputStream()) {
                override fun close() {
                    out.close()
                    val bos = out as ByteArrayOutputStream
                    classBytes!!.put(className, bos.toByteArray())
                }
            }
        }
    }

    companion object {

        private val JAVA_SOURCE_FILE_EXT = ".java"

        fun toUri(name: String): URI {
            val newUri = StringBuilder()
            newUri.append("mfm:///${name.replace('.', '/')}")

            if (name.endsWith(JAVA_SOURCE_FILE_EXT)) {
                newUri.replace(newUri.length - JAVA_SOURCE_FILE_EXT.length, newUri.length, JAVA_SOURCE_FILE_EXT)
            }

            return URI.create(newUri.toString())
        }
    }
}
